File: mechanical/src/main/java/resonantinduction/mechanical/process/purifier/TileMixer.java
package resonantinduction.mechanical.process.purifier;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

import net.minecraft.block.Block;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.item.ItemStack;
import net.minecraft.util.AxisAlignedBB;
import net.minecraftforge.common.ForgeDirection;
import resonantinduction.api.mechanical.IMechanical;
import resonantinduction.api.recipe.MachineRecipes;
import resonantinduction.api.recipe.MachineRecipes.RecipeType;
import resonantinduction.core.Reference;
import resonantinduction.core.ResonantInduction;
import resonantinduction.core.resource.ResourceGenerator;
import resonantinduction.core.resource.fluid.BlockFluidMixture;
import resonantinduction.mechanical.network.TileMechanical;
import resonantinduction.mechanical.process.Timer;
import universalelectricity.api.vector.Vector3;

/**
 * @author Calclavia
 * 
 */
public class TileMixer extends TileMechanical
{
	public static final long POWER = 500000;
	public static final int PROCESS_TIME = 5 * 20;
	public static final Timer<EntityItem> timer = new Timer<EntityItem>();

	@Override
	public void updateEntity()
	{
		super.updateEntity();

		if (canWork())
		{
			doWork();
		}
	}

	/**
	 * Can this machine work this tick?
	 * 
	 * @return
	 */
	public boolean canWork()
	{
		return angularVelocity != 0;
	}

	public void doWork()
	{
		boolean didWork = false;

		// Search for an item to "process"
		AxisAlignedBB aabb = AxisAlignedBB.getAABBPool().getAABB(this.xCoord - 1, this.yCoord, this.zCoord - 1, this.xCoord + 2, this.yCoord + 1, this.zCoord + 2);
		List<Entity> entities = this.worldObj.getEntitiesWithinAABB(Entity.class, aabb);
		Set<EntityItem> processItems = new LinkedHashSet<EntityItem>();

		for (Entity entity : entities)
		{
			/**
			 * Rotate entities around the mixer
			 */
			Vector3 originalPosition = new Vector3(entity);
			Vector3 relativePosition = originalPosition.clone().subtract(new Vector3(this).add(0.5));
			relativePosition.rotate(-angularVelocity, 0, 0);
			Vector3 newPosition = new Vector3(this).add(0.5).add(relativePosition);
			Vector3 difference = newPosition.difference(originalPosition).scale(0.5);

			entity.addVelocity(difference.x, difference.y, difference.z);
			entity.onGround = false;

			if (entity instanceof EntityItem)
			{
				if (MachineRecipes.INSTANCE.getOutput(RecipeType.MIXER, ((EntityItem) entity).getEntityItem()).length > 0)
				{
					processItems.add((EntityItem) entity);
				}
			}
		}

		for (EntityItem processingItem : processItems)
		{
			if (!timer.containsKey(processingItem))
			{
				timer.put(processingItem, PROCESS_TIME);
			}

			if (!processingItem.isDead && new Vector3(this).add(0.5).distance(processingItem) < 2)
			{
				int timeLeft = timer.decrease(processingItem);

				if (timeLeft <= 0)
				{
					if (this.doneWork(processingItem))
					{
						if (--processingItem.getEntityItem().stackSize <= 0)
						{
							processingItem.setDead();
							timer.remove(processingItem);
							processingItem = null;
						}
						else
						{
							processingItem.setEntityItemStack(processingItem.getEntityItem());
							// Reset timer
							timer.put(processingItem, PROCESS_TIME);
						}
					}
				}
				else
				{
					processingItem.delayBeforeCanPickup = 20;
					this.worldObj.spawnParticle("bubble", processingItem.posX, processingItem.posY, processingItem.posZ, (Math.random() - 0.5f) * 3, (Math.random() - 0.5f) * 3, (Math.random() - 0.5f) * 3);
				}

				didWork = true;
			}
			else
			{
				timer.remove(processingItem);
				processingItem = null;
			}
		}

		if (didWork)
		{
			if (this.ticks % 20 == 0)
			{
				this.worldObj.playSoundEffect(this.xCoord + 0.5, this.yCoord + 0.5, this.zCoord + 0.5, Reference.PREFIX + "mixer", 0.5f, 1);
			}
		}
	}

	private boolean doneWork(EntityItem entity)
	{
		Vector3 mixPosition = new Vector3(entity.posX, yCoord, entity.posZ);
		Block block = Block.blocksList[mixPosition.getBlockID(worldObj)];

		if (block instanceof BlockFluidMixture)
		{
			ItemStack itemStack = entity.getEntityItem().copy();

			if (((BlockFluidMixture) block).mix(worldObj, mixPosition.intX(), mixPosition.intY(), mixPosition.intZ(), itemStack))
			{
				worldObj.notifyBlocksOfNeighborChange(mixPosition.intX(), mixPosition.intY(), mixPosition.intZ(), mixPosition.getBlockID(worldObj));
				return true;
			}
		}
		else if (worldObj.isAirBlock(mixPosition.intX(), mixPosition.intY(), mixPosition.intZ()) || block.blockID == Block.waterStill.blockID || block.blockID == Block.waterMoving.blockID)
		{
			mixPosition.setBlock(worldObj, ResourceGenerator.getMixture(ResourceGenerator.getName(entity.getEntityItem().getItemDamage())).blockID);
		}

		return false;
	}

	@Override
	public boolean canConnect(ForgeDirection from, Object source)
	{
		return from == ForgeDirection.UP || from == ForgeDirection.DOWN;
	}

	@Override
	public boolean inverseRotation(ForgeDirection dir, IMechanical with)
	{
		return false;
	}
}

File: src/main/scala/edx/mechanical/mech/process/mixer/TileMixer.scala
package edx.mechanical.mech.process.mixer

import java.util.{LinkedHashSet, List, Set}

import cpw.mods.fml.relauncher.{Side, SideOnly}
import edx.core.Reference
import edx.mechanical.mech.TileMechanical
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.entity.Entity
import net.minecraft.entity.item.EntityItem
import net.minecraft.item.ItemStack
import net.minecraft.util.{AxisAlignedBB, ResourceLocation}
import net.minecraftforge.client.model.{AdvancedModelLoader, IModelCustom}
import net.minecraftforge.fluids.IFluidBlock
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL11._
import resonantengine.api.recipe.{MachineRecipes, RecipeType}
import resonantengine.lib.prefab.Timer
import resonantengine.lib.render.RenderUtility
import resonantengine.lib.transform.rotation.Quaternion
import resonantengine.lib.transform.vector.Vector3

import scala.collection.JavaConversions._

/**
 * @author Calclavia
 */
object TileMixer
{
  final val PROCESS_TIME: Int = 12 * 20
  final val MIXER_ITEM_TIMER: Timer[EntityItem] = new Timer[EntityItem]
  @SideOnly(Side.CLIENT) val MODEL: IModelCustom = AdvancedModelLoader.loadModel(new ResourceLocation(Reference.domain, Reference.modelPath + "mixer.tcn"))
  @SideOnly(Side.CLIENT) var TEXTURE: ResourceLocation = new ResourceLocation(Reference.domain, Reference.modelPath + "mixer.png")
}

class TileMixer extends TileMechanical(Material.iron)
{
  private var areaBlockedFromMoving: Boolean = false

  //Constructor
  mechanicalNode = new MixerNode(this)
  isOpaqueCube = false
  normalRender = false
  customItemRender = true
  textureName = "material_metal_top"

  override def update
  {
    super.update
    if (!world.isRemote && ticks % 3 == 0)
    {
      this.areaBlockedFromMoving = checkIsBlocked
    }
    if (canWork)
    {
      doWork
    }
  }

  /** Checks to see if the area around the mixer is blocked (3x3 - excluding center)
    * @return true if there is a non-fluid block inside the bounds
    */
  def checkIsBlocked: Boolean =
  {
    for (x <- -1 to 1)
    {
      for (z <- -1 to 1)
      {
        if (x != 0 && z != 0)
        {
          val block: Block = toVector3.add(x, 0, z).getBlock(world)
          if (block != null && !(block.isInstanceOf[IFluidBlock]))
          {
            return true
          }
        }
      }
    }
    return false
  }

  /**
   * Can this machine work this tick?
   *
   * @return
   */
  def canWork: Boolean =
  {
    return mechanicalNode.angularVelocity != 0 && !areaBlockedFromMoving
  }

  def doWork
  {
    var didWork: Boolean = false
    val aabb: AxisAlignedBB = AxisAlignedBB.getBoundingBox(this.xCoord - 1, this.yCoord, this.zCoord - 1, this.xCoord + 2, this.yCoord + 1, this.zCoord + 2)
    val entities: List[_] = this.worldObj.getEntitiesWithinAABB(classOf[Entity], aabb)
    val processItems: Set[EntityItem] = new LinkedHashSet[EntityItem]
    for (obj <- entities)
    {
      val entity: Entity = obj.asInstanceOf[Entity]
      val originalPosition: Vector3 = new Vector3(entity)
      val relativePosition: Vector3 = originalPosition - toVector3.add(0.5)
      relativePosition.transform(new Quaternion(-mechanicalNode.angularVelocity, new Vector3(1, 0, 0)))
      val newPosition: Vector3 = toVector3 + 0.5 + relativePosition
      val difference: Vector3 = (newPosition - originalPosition) * 0.5
      entity.addVelocity(difference.x, difference.y, difference.z)
      entity.onGround = false
      if (entity.isInstanceOf[EntityItem])
      {
        if (MachineRecipes.instance.getOutput(RecipeType.MIXER.name, (entity.asInstanceOf[EntityItem]).getEntityItem).length > 0)
        {
          processItems.add(entity.asInstanceOf[EntityItem])
        }
      }
    }

    for (processingItem <- processItems)
    {
      if (!TileMixer.MIXER_ITEM_TIMER.containsKey(processingItem))
      {
        TileMixer.MIXER_ITEM_TIMER.put(processingItem, TileMixer.PROCESS_TIME)
      }
      if (!processingItem.isDead && (toVector3 + 0.5).distance(new Vector3(processingItem)) < 2)
      {
        val timeLeft: Int = TileMixer.MIXER_ITEM_TIMER.decrease(processingItem)
        if (timeLeft <= 0)
        {
          if (doneWork(processingItem))
          {
            processingItem.getEntityItem.stackSize -= 1

            if (processingItem.getEntityItem.stackSize <= 0)
            {
              processingItem.setDead
              TileMixer.MIXER_ITEM_TIMER.remove(processingItem)
            }
            else
            {
              processingItem.setEntityItemStack(processingItem.getEntityItem)
              TileMixer.MIXER_ITEM_TIMER.put(processingItem, TileMixer.PROCESS_TIME)
            }
          }
        }
        else
        {
          processingItem.delayBeforeCanPickup = 20
          this.worldObj.spawnParticle("bubble", processingItem.posX, processingItem.posY, processingItem.posZ, (Math.random - 0.5f) * 3, (Math.random - 0.5f) * 3, (Math.random - 0.5f) * 3)
        }
        didWork = true
      }
      else
      {
        TileMixer.MIXER_ITEM_TIMER.remove(processingItem)
      }
    }
    if (didWork)
    {
      if (this.ticks % 20 == 0)
      {
        this.worldObj.playSoundEffect(this.xCoord + 0.5, this.yCoord + 0.5, this.zCoord + 0.5, Reference.prefix + "mixer", 0.5f, 1)
      }
    }
  }

  private def doneWork(entity: EntityItem): Boolean =
  {
    val mixPosition: Vector3 = new Vector3(entity.posX, yCoord, entity.posZ)
    if (mixPosition.getBlock(world) ne getBlockType)
    {
      val block: Block = mixPosition.getBlock(worldObj)
      val blockFluidFinite: Block = null //RIResourceFactory.getMixture(ResonantEngine.resourceFactory.getName(entity.getEntityItem))
      if (blockFluidFinite != null)
      {
        /*if (block.isInstanceOf[BlockFluidMixture])
        {
          val itemStack: ItemStack = entity.getEntityItem.copy
          if (block.asInstanceOf[BlockFluidMixture].mix(worldObj, mixPosition.xi, mixPosition.yi, mixPosition.zi, itemStack))
          {
            worldObj.notifyBlocksOfNeighborChange(mixPosition.xi, mixPosition.yi, mixPosition.zi, mixPosition.getBlock(worldObj))
            return true
          }
        }
        else if (block != null && (block == Blocks.water || block == Blocks.flowing_water))
        {
          mixPosition.setBlock(worldObj, blockFluidFinite)
        }*/
      }
    }
    return false
  }

  override def renderDynamic(position: Vector3, frame: Float, pass: Int): Unit =
  {
    GL11.glPushMatrix()
    GL11.glTranslated(position.x + 0.5, position.y + 0.5, position.z + 0.5)
    RenderUtility.bind(RenderMixer.TEXTURE)
    RenderMixer.MODEL.renderOnly("centerTop", "centerBase")
    glPushMatrix()
    glRotatef(Math.toDegrees(mechanicalNode.angle.asInstanceOf[Float]).asInstanceOf[Float], 0, 1, 0)
    RenderMixer.MODEL.renderAllExcept("centerTop", "centerBase")
    glPopMatrix()
    GL11.glPopMatrix()
  }

  override def renderInventory(itemStack: ItemStack)
  {
    glPushMatrix()
    GL11.glScalef(0.7f, 0.7f, 0.7f)
    glTranslatef(0.5F, 0.5f, 0.5f)
    RenderUtility.bind(RenderMixer.TEXTURE)
    RenderMixer.MODEL.renderAll()
    glPopMatrix()
  }
}